heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java

```diff
@@ -0,0 +1,384 @@
+
+    this.instanceRamDefault = Context.instanceRam(config);
+    this.instanceCpuDefault = Context.instanceCpu(config).doubleValue();
+    this.instanceDiskDefault = Context.instanceDisk(config);
+  }
+
+  @Override
+  public PackingPlan pack() {
+    // Get the instances using FFD allocation
+    Map<String, List<String>> ffdAllocation = getFFDAllocation();
+
+    // Construct the PackingPlan
+    Map<String, PackingPlan.ContainerPlan> containerPlanMap = new HashMap<>();
+    Map<String, Long> ramMap = TopologyUtils.getComponentRamMapConfig(topology);
+
+    long topologyRam = 0;
+    long topologyDisk = 0;
+    double topologyCpu = 0.0;
+
+    for (Map.Entry<String, List<String>> entry : ffdAllocation.entrySet()) {
+
+      String containerId = entry.getKey();
+      List<String> instanceList = entry.getValue();
+
+      long containerRam = DEFAULT_RAM_PADDING_PER_CONTAINER;
+      long containerDiskInBytes = DEFAULT_DISK_PADDING_PER_CONTAINER;
+      double containerCpu = DEFAULT_CPU_PADDING_PER_CONTAINER;
+
+      // Calculate the resource required for single instance
+      Map<String, PackingPlan.InstancePlan> instancePlanMap = new HashMap<>();
+
+      for (String instanceId : instanceList) {
+        long instanceRam = 0;
+        if (ramMap.containsKey(getComponentName(instanceId))) {
+          instanceRam = ramMap.get(getComponentName(instanceId));
+          if (!isValidInstance(instanceRam)) {
+            return null;
+          }
+        } else {
+          instanceRam = instanceRamDefault;
+        }
+        containerRam += instanceRam;
+
+        // Currently not yet support disk or cpu config for different components,
+        // so just use the default value.
+        long instanceDisk = instanceDiskDefault;
+        containerDiskInBytes += instanceDisk;
+
+        double instanceCpu = instanceCpuDefault;
+        containerCpu += instanceCpu;
+
+        PackingPlan.Resource resource =
+            new PackingPlan.Resource(instanceCpu, instanceRam, instanceDisk);
+        PackingPlan.InstancePlan instancePlan =
+            new PackingPlan.InstancePlan(
+                instanceId,
+                getComponentName(instanceId),
+                resource);
+        // Insert it into the map
+        instancePlanMap.put(instanceId, instancePlan);
+      }
+
+      PackingPlan.Resource resource =
+          new PackingPlan.Resource(containerCpu, containerRam, containerDiskInBytes);
+      PackingPlan.ContainerPlan containerPlan =
+          new PackingPlan.ContainerPlan(containerId, instancePlanMap, resource);
+
+      containerPlanMap.put(containerId, containerPlan);
+      topologyRam += containerRam;
+      topologyCpu += containerCpu;
+      topologyDisk += containerDiskInBytes;
+    }
+
+    // Take the heron internal container into account and the application master for YARN
+    // scheduler
+    topologyRam += 2 * (instanceRamDefault + DEFAULT_RAM_PADDING_PER_CONTAINER);
```

Comment:
Good catch. Each topology running on YARN needs one extra container for application master. However the resources allocated to AM container and TMaster container are constant and independent of instanceRam and default padding. Which means the total resources allocated to a topology is a `constant + cost-of-deploying-instances` for both `yarn` and `aurora` clusters.